File: GFG POTD/023-October-2025/017-bst-to-greater-sum-tree.cpp
// https://www.geeksforgeeks.org/problems/bst-to-greater-sum-tree/1/




#include <bits/stdc++.h>
using namespace std;

class Node {
  public:
    int data;
    Node* left;
    Node* right;

    Node(int value) {
        data = value;
        left = nullptr;
        right = nullptr;
    }
};

class Solution {
  private:
    void solve(Node *root, int &sum) {
        if (!root) return;
        
        solve(root -> right, sum);
        
        int temp = root -> data;
        root -> data = sum;
        sum += temp;
        
        solve(root -> left, sum);
    }
  public:
    // TC : O(N)
    // SC : O(1)
    void transformTree(Node *root) {
        int sum = 0;
        solve(root, sum);
    }
};

File: GFG POTD/023-October-2025/017-bst-to-greater-sum-tree.java
// https://www.geeksforgeeks.org/problems/bst-to-greater-sum-tree/1/




class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data = data;
        left=null;
        right=null;
    }
}

class Solution {
    private static int sum = 0;
    
    // TC : O(N)
    // SC : O(1)
    public static void transformTree(Node root) {
        sum = 0;
        solve(root);
    }
    
    private static void solve(Node root) {
        if (root == null) return;
        
        solve(root.right);
        
        int temp = root.data;
        root.data = sum;
        sum += temp;
        
        solve(root.left);
    }
}